File: tests/chromium_smoke.py
"""Chromium verification against the live preview (Playwright)."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from playwright.sync_api import sync_playwright

OUT = Path(__file__).parent / "reports"
OUT.mkdir(exist_ok=True)
SHOT = Path("/workspace/screenshots")
SHOT.mkdir(exist_ok=True)


def main() -> int:
    url = sys.argv[1] if len(sys.argv) > 1 else "http://127.0.0.1:8080/"
    cases = []
    console_errors = []
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(viewport={"width": 1280, "height": 800})
        page.on("pageerror", lambda err: console_errors.append(str(err)))
        page.on(
            "console",
            lambda msg: console_errors.append(msg.text) if msg.type == "error" else None,
        )
        page.goto(url, wait_until="networkidle", timeout=30000)
        body = page.inner_text("body")
        cases.append({"name": "home renders Forge Layer", "ok": "Forge Layer" in body, "detail": body[:80]})
        page.screenshot(path=str(SHOT / "chromium-home.png"))

        page.click("text=Registry")
        page.wait_for_timeout(800)
        body = page.inner_text("body")
        cases.append({"name": "registry lists dockets", "ok": "FL-" in body, "detail": "docket ids present"})
        page.screenshot(path=str(SHOT / "chromium-registry.png"))

        page.goto(url + "submit", wait_until="networkidle")
        body = page.inner_text("body")
        cases.append({"name": "submit form", "ok": "File a dispute" in body, "detail": "form visible"})

        page.goto(url + "protocol", wait_until="networkidle")
        body = page.inner_text("body")
        cases.append({"name": "protocol page", "ok": "Equivalence" in body or "validators" in body.lower(), "detail": "mechanism copy"})

        page.set_viewport_size({"width": 390, "height": 844})
        page.goto(url, wait_until="networkidle")
        overflow = page.evaluate("() => document.documentElement.scrollWidth > document.documentElement.clientWidth + 1")
        cases.append({"name": "mobile no overflow", "ok": not overflow, "detail": str(overflow)})

        failed_net = []
        page.on("requestfailed", lambda req: failed_net.append(req.url))
        browser.close()

    unexpected = [e for e in console_errors if "aborted" not in e.lower()]
    cases.append({"name": "no unexpected console errors", "ok": len(unexpected) == 0, "detail": str(unexpected[:5])})
    status = "PASS" if all(c["ok"] for c in cases) else "FAIL"
    payload = {"status": status, "cases": cases, "console": unexpected}
    (OUT / "09_chromium.json").write_text(json.dumps(payload, indent=2))
    lines = [f"# Chromium Report", "", f"Status: **{status}**", "", "## Cases", ""]
    for c in cases:
        lines.append(f"- [{'PASS' if c['ok'] else 'FAIL'}] {c['name']}: {c['detail']}")
    (OUT / "09_chromium.md").write_text("\n".join(lines) + "\n")
    print(json.dumps(payload, indent=2))
    return 0 if status == "PASS" else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/chromium_smoke.mjs
#!/usr/bin/env node
/**
 * Chromium verification against the QA stack (or any base URL).
 * Uses the workspace Playwright install.
 */
import { chromium } from "playwright";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const REPORTS = path.join(HERE, "reports");
const SHOT = "/workspace/screenshots";
fs.mkdirSync(REPORTS, { recursive: true });
fs.mkdirSync(SHOT, { recursive: true });

const url = process.argv[2] || "http://127.0.0.1:8080/";
const base = url.endsWith("/") ? url : url + "/";

const cases = [];
const consoleErrors = [];
const failedNet = [];
const notFound = [];

function add(name, ok, detail) {
  cases.push({ name, ok, detail: String(detail ?? "") });
}

const browser = await chromium.launch({ headless: true });
const page = await browser.newPage({ viewport: { width: 1280, height: 800 } });
page.on("pageerror", (err) => consoleErrors.push(String(err)));
page.on("console", (msg) => {
  if (msg.type() === "error") consoleErrors.push(msg.text());
});
page.on("requestfailed", (req) => failedNet.push(`${req.method()} ${req.url()} ${req.failure()?.errorText || ""}`));
page.on("response", (res) => {
  if (res.status() === 404) notFound.push(res.url());
});

await page.goto(base, { waitUntil: "networkidle", timeout: 30000 });
await page.waitForTimeout(600);
let body = await page.innerText("body");
add("home renders Forge Layer", body.includes("Forge Layer"), body.slice(0, 120));
add("home shows contract-not-configured or live chip", /Contract not configured|Local rehearsal|studionet|testnet/i.test(body), "banner/chip");
add("landing has citation specimen", /What a citation looks like|A closed record/i.test(body), "specimen");
add("landing has trust properties", /Equivalence Principle|Wallet-signed writes/i.test(body), "holds");
add("landing has roadmap rail", /The forge|On the record/i.test(body), "roadmap");
await page.screenshot({ path: path.join(SHOT, "chromium-home.png") });

await page.click("text=Registry");
await page.waitForTimeout(900);
body = await page.innerText("body");
add("registry lists dockets", body.includes("FL-") || body.includes("Registry"), "docket ids or registry heading");
await page.screenshot({ path: path.join(SHOT, "chromium-registry.png") });

await page.goto(base + "submit", { waitUntil: "networkidle" });
await page.waitForTimeout(400);
body = await page.innerText("body");
add("submit form", body.includes("File a dispute"), "form visible");
await page.screenshot({ path: path.join(SHOT, "chromium-submit.png") });

await page.goto(base + "roadmap", { waitUntil: "networkidle" });
body = await page.innerText("body");
add("roadmap page", /What we ship|Chamber|Attestation/i.test(body), "roadmap copy");

await page.goto(base + "protocol", { waitUntil: "networkidle" });
body = await page.innerText("body");
add("protocol page", /Equivalence|validators/i.test(body), "mechanism copy");

await page.goto(base + "disputes/1/", { waitUntil: "networkidle" });
await page.waitForTimeout(800);
body = await page.innerText("body");
add("docket detail centerpiece", /FL-00001/.test(body) && /Cited work|Consensus|Challenge/i.test(body), body.slice(0, 80));
await page.screenshot({ path: path.join(SHOT, "chromium-docket.png") });

await page.setViewportSize({ width: 390, height: 844 });
await page.goto(base, { waitUntil: "networkidle" });
await page.waitForTimeout(400);
const overflow = await page.evaluate(
  () => document.documentElement.scrollWidth > document.documentElement.clientWidth + 1,
);
add("mobile no overflow", !overflow, String(overflow));
await page.screenshot({ path: path.join(SHOT, "chromium-mobile.png") });

await browser.close();

const origin = new URL(base).origin;
const ignorable404 = (u) => {
  try {
    const parsed = new URL(u);
    if (parsed.origin !== origin) return true; // cited third-party content
  } catch {
    return true;
  }
  return /favicon\.ico|apple-touch-icon|manifest\.json|\.map$/i.test(u);
};
const unexpected404 = notFound.filter((u) => !ignorable404(u));
const unexpected = consoleErrors.filter((e) => {
  if (/aborted|favicon|net::ERR_ABORTED/i.test(e)) return false;
  if (/Failed to load resource: the server responded with a status of 404/i.test(e) && unexpected404.length === 0) {
    return false;
  }
  return true;
});
add("no unexpected console errors", unexpected.length === 0, JSON.stringify(unexpected.slice(0, 5)));
add(
  "no failed network on core paths",
  failedNet.filter((u) => !/favicon|sourcemap/i.test(u)).length === 0 && unexpected404.length === 0,
  JSON.stringify({ failedNet: failedNet.slice(0, 5), notFound: unexpected404.slice(0, 8) }),
);

const status = cases.every((c) => c.ok) ? "PASS" : "FAIL";
const payload = { status, cases, console: unexpected, failedNet, url: base };
fs.writeFileSync(path.join(REPORTS, "09_chromium.json"), JSON.stringify(payload, null, 2));
const lines = ["# Chromium Report", "", `Status: **${status}**`, "", `URL: ${base}`, "", "## Cases", ""];
for (const c of cases) {
  lines.push(`- [${c.ok ? "PASS" : "FAIL"}] ${c.name}: ${c.detail}`);
}
fs.writeFileSync(path.join(REPORTS, "09_chromium.md"), lines.join("\n") + "\n");
console.log(JSON.stringify(payload, null, 2));
process.exit(status === "PASS" ? 0 : 1);
